src/jarabe/util/telepathy/connection_watcher.py

```diff
@@ -56,8 +60,12 @@ def __init__(self, bus=None):
                                      signal_name='StatusChanged',
                                      path_keyword='path')
 
-        for conn in Connection.get_connections(bus):
-            conn.call_when_ready(self._conn_ready_cb)
+        bus_object = self.bus.get_object('org.freedesktop.DBus', '/org/freedesktop/DBus')
+        for service in bus_object.ListNames(dbus_interface='org.freedesktop.DBus'):
```

Comment:
This is a synchronous call to D-Bus; can it be made asynchronous?

This method of asking D-Bus to list names, only to locate names that represent connections, causes traffic on D-Bus.  Could we instead keep a list of connections and iterate through that?

Reply:
Thanks. Made a list of connections, to iterate through.

Reply:
No, that's not what I meant.  I'll try again.

In _master_ the ConnectionWatcher calls get_connections, a method of the Connection class.  In client/conn.py of telepathy-python sources, this is revealed to be a synchronous call to ListNames.  Each detected name is then used to create a new Connection.

Synchronous calls are to be avoided because they can force interprocess context switches, clearing processor caches, and delay response by the application that makes the call.  So I'm asking if we can avoid synchronous calls by either avoiding the ListNames call altogether (by caching our own list of connections), or if we are not in command of the list of connections (e.g. because they can be created by other processes), at least make the call to ListNames asynchronous, which would require changes to respond to the completion event.

However, if you can say "ListNames completes without requiring action by other processes", I'd say leave it synchronous.

Is there an equivalent feature in TelepathyGLib that we can look at to understand, though not to use?

Reply:
> or if we are not in command of the list of connections
Yes we are not in command of that I have checked that with developing list of services with and without sugar. Also the same service have unique id's for different users.
```
[dbus.String(u'org.freedesktop.Telepathy.Connection.salut.local_xmpp.e26599a6') 
```
The 8 digit hash at the last differ for different users.

Reply:
> No, that's not what I meant. I'll try again.

Thanks. I think I understood your point this time. What we want is to make the synchronous `LIstNames` call an asynchronous call. I have pushed https://github.com/sugarlabs/sugar/pull/837/commits/0b9bcef7f8d016aaf06883b9f517a47cc6330a68 in favor of that, please review.